### ISSUE TITLE
Replace DirtyNIF execution model with threadpool

### ISSUE THREAD

**Issue by warmwaffles** [elixir-sqlite/exqlite#192]
The dirtynif execution model is okay for most cases, but we have an issue where long running writes need to timeout / be interrupted mid execution and aborted. The only way to do that is to utilize https://sqlite.org/c3ref/progress_handler.html

**Comment by josevalim** [comment]
I wonder if you may be able to stick with the process model as long as you assume each instance of the database is backed by one process. This way you make Elixir processes your thread pool, instead of having to deal with it in C/C++/Rust.

**Comment by warmwaffles** [comment]
@josevalim I think that's a good idea where only one process in the supervision tree sits there waiting for another `DbConnection` call. This _could_ make it easier to cancel existing running queries.

Let me noodle on this a little more.

**Comment by josevalim** [comment]
Right, I think in this sense DBConnection may be getting more in the way than helping. Because you need serial access around SQLite3 and having a pool around serial access is not giving you anything. And when you don't need serial access (i.e. read operations), you don't need a pool at all.

DBConnection is necessary for Ecto but, even then, it should not be a requirement. There are only three features in Ecto.Adapters.SQL that require DBConnection (sandbox, stream, and disconnect_all) but they are all addressable.

Then there is the part of DBConnection as a convenience. I can think of two features it gives:

1. a transaction manager - if the process doing a transaction crashes, we need to abort it. but perhaps sqlite3 api itself already has conveniences for handling dangling transactions
2. the ownership pool

Of course there is a lot of work here, so I am not saying we should yank DBConnection, but it may be worth considering moving the DBConnection bits to ecto_sqlite3. In this project, based on the analysis above (which may be incomplete), you would only be missing the transaction manager (and it may be provided by sqlite3).

**Comment by LostKobrakai** [comment]
There might be an additional benefit to managing transactions in a sqlite specific way. Currently there's no information returned to changesets about which constraint was violated, but there seems to be some pragmas to figure those out as sqlite doesn't close transactions on errors.

What I'm wondering about @josevalim's suggestion is how using processes to serialize or not serialize access would work for transactions which only become write transactions mid execution. Wouldn't the suggested approach require knowing if a transaction does writes or not in advance?

**Comment by ruslandoga** [comment]
Wouldn't having one process per db instance negate WAL benefits? There still needs to be a pool of connections for concurrent reads. It's not possible to open a single connection and start multiple transactions on it.

**Comment by josevalim** [comment]
@ruslandoga the reads can go directly to the NIF. Only the writes and operations inside a transaction would go through the process.

**Comment by ruslandoga** [comment]
But how exactly reads would go directly to the NIF? Using what database connection?

**Comment by LostKobrakai** [comment]
I think the tricky thing would be supporting concurrent reads from within transactions, which didn't yet do writes:
https://www.sqlite.org/lang_transaction.html#deferred_immediate_and_exclusive_transactions

Not sure how exqlite could know which statement within a transaction requires a SHARED LOCK vs one of the ones required for writing.

**Comment by LostKobrakai** [comment]
I just tried this with two table plus instances on the same database and it allowed me to start a read transaction just fine while another write transaction was started. Trying to start a write transaction I got  `database is locked`. Though I guess that's two "connections" to sql, is it?

**Comment by ruslandoga** [comment]
>  two table plus instances 

That means two connections. I think the approach discussed in this issue is having only a single connection open to the database.

**Comment by LostKobrakai** [comment]
Given DBConnection is an abstraction over connections I'm not so sure. I'd for sure appreciate if it actually could do concurrent reads (in a transaction) given how much code in a business application requires to be run in a transaction, which usually start reading a bunch of things before they write.

**Comment by ruslandoga** [comment]
I think there is danger of those transactions failing with `SQLITE_BUSY` so serialising all transactions seems like a safer way.

**Comment by LostKobrakai** [comment]
That's why I was wondering if exqlite could catch writes before they queue for "BUSY WAIT", but the queuing happens in elixir land. In the end this wouldn't be much different to e.g. a timeout due to a to long running query with e.g. postgres/ecto. Another option would be possible if exqlite can timeout a transaction from the outside and setting busy_timeout to a value larger than elixirs internal timeout.

**Comment by ruslandoga** [comment]
I don't understand. How would busy_timeout help in the above example? No matter how long we wait, if we started a deferred transaction before some other connection made a write, we won't be able to commit it even if the writer commits first.

**Comment by LostKobrakai** [comment]
That's unexpected, I would've expected this to work based on the documentation on the topic. If this is the case I'd wonder why have busy_timeout in the first place, if waiting makes no sense.

**Comment by ruslandoga** [comment]
What if we try updating a record we read but that since has been deleted? I don't think it can work at all.

**Comment by LostKobrakai** [comment]
It there are conflicting changes I think most databases will roll back one transaction. Though deleting one row and adding another shouldn't really be conflicting, but that might just be sqlite.

**Comment by ruslandoga** [comment]
I'm back to thinking that a single connection with un-pooled reads wouldn't work. If we issue reads directly to the NIF while a writer is in a transaction, we lose isolation guarantees as we can read data from unfinished transaction.

```elixir
iex> spawn(fn ->
  Exqlite.transaction(db, fn conn ->
    Exqlite.query(conn, "insert into users(name) values(?)", ["new"])
    :timer.sleep(:timer.seconds(1000))
  end)
end)
iex> Exqlite.query(db, "select * from users")
{:ok,
 [
   ["david"],
   ["peter"],
   ["new"]
 ]}
```

**Comment by ruslandoga** [comment]
> I'm thinking now about using two connections, one for writers and one for readers. The readers connection would accept un-pooled reads.

That didn't end up working either. Reads start an implicit transaction and that's a problem since concurrent reads can force that transaction to be indefinitely open and that connection would never get new data.

**Comment by josevalim** [comment]
Sorry for staying out of this conversation and thanks for exploring @ruslandoga.

It seems like we would need the process to work as read-write lock. While reads are happening, we allow them to run concurrently. Once a write arrives, we start queueing events and draining the reads. Once they are all drained, execute the write, and then move to the next in the queue. If a read, we start doing concurrent work again. All queries can still run on the client. You can find an implementation of serialized reads-writes here: https://github.com/elixir-explorer/adbc/blob/main/lib/adbc_connection.ex - having a concurrent reads version should an expansion of that API.

API wise, you could keep "query" as a write operation but you can introduce a "read_query" where the user guarantees they are exclusively reading.

In any case, the main point is that I think this library can be cleaner and easier to maintain if it doesn't try to do NIFs, process management, and DBConnection at the same time. I think NIFs + process management would make this usable across more scenarios and DBConnection is an Ecto concern (which perhaps we can move all the way up to Ecto itself).

**Comment by ruslandoga** [comment]
@josevalim @warmwaffles if there no other takers, I'm interested in exploring this approach :)

**Comment by josevalim** [comment]
Definitely go for it from my side - but that's @warmwaffles decision to make. :)

**Comment by warmwaffles** [comment]
Yea sure why not. @ruslandoga if you want to give it a shot I'm all for it.

**Comment by ruslandoga** [comment]
I've started a PR https://github.com/elixir-sqlite/exqlite/pull/255 and while working on it I had one concern. With this approach we miss out on WAL allowing reads during writes, but I guess it could be in `ecto_sqlite3` by starting two connections and, for example, dedicating one for writes and the other for reads. After the write happens, the read connection is drained from reads to close the implicit transaction and allow for the new data to become visible.

**Comment by josevalim** [comment]
Before I can answer, quick question: can we perform concurrent reads over the same single connection in SQLite3? Or the reads have to be serialized?

**Comment by ruslandoga** [comment]
Yes, we can perform concurrent reads, the "result rows" are "scoped" per statement. So as we are stepping over two different prepared statements in the same SQLite connection from different Elixir processed, we don't get interleaved results and all is fine :)

**Comment by warmwaffles** [comment]
SQLite AFAIK can handle concurrent reads, but the moment a write comes along, things get tricky. It also depends on how the database's `journal_mode` is set. If it is set to `wal`, sqlite will do its best to resolve the writes that happen concurrently.

**Comment by ruslandoga** [comment]
> So my suggestion is to have the RWLock start both connections (we could make the Read transaction be lazy in the future if that's ever an issue). query goes through the write one, read_query goes through the read one.

I like this approach: https://github.com/elixir-sqlite/exqlite/pull/256

My only concern is that it seems a bit sneaky. I'm trying to come up with scenarios where it can bite the user, for example `:memory:` databases, but I'll handle that next.

**Comment by warmwaffles** [comment]
@ruslandoga another interesting thought I had is that ideally the database is read from more than written to. Bottlenecks will be at the two genservers spun up, mainly on the reader. We _could_ try to pool the readers and round robin the calls to each one, and maintain a single write instance. Probably something to explore later.

**Comment by ruslandoga** [comment]
The only thing the reader does in "normal" state is preparing the statements. Since there is no cache, we can move that to the caller as well. Then the bottleneck would only be able to appear during read connection draining after a locked query finishes (assuming it was a write). During this state it might be possible to get a long read queue.

**Comment by warmwaffles** [comment]
One of the main reasons I wanted to look into replacing the DirtyNIF execution model is that I would like to be able to cancel queries that overrun the timeout. Ecto would timeout requests if they took too long to run while modifying records.

One use case was a developer I was talking to was updating / cleaning up records and the query was touching thousands of records overrunning the default timeout. Although it looked like it timed out, instead it was still running in the background and causing compounding latency issues.

What my hope is that with the new implementation that you are working on @ruslandoga we are not going to timeout the connection at all, but instead keep it open until the request finishes.

**Comment by ruslandoga** [comment]
Right now I'm exploring the following approach:

- exqlite becomes just the NIF wrapper, like https://github.com/ruslandoga/sqlite.ex, but exposes some more utility functions
- ecto_sqlite3 starts two processes, a reader and a writer (for a readwrite db), with reader multiplexing select statements

I'll try to finish it up by the end of the week and open new PRs.

**Comment by ruslandoga** [comment]
> One of the main reasons I wanted to look into replacing the DirtyNIF execution model is that I would like to be able to cancel queries that overrun the timeout. Ecto would timeout requests if they took too long to run while modifying records.

Without DBConnection there wouldn't be any timeouts, I think. But if we want to still support them, it can be done both with DirtyNIFs and a slightly different INSERT handling: instead of building a single big multi-row `INSERT INTO ... (?, ?, ?), (?, ?, ?), (?, ?, ?)` statement we can prepare an insert for a single row and re-execute it for every row. We can do it similar to `multi_step` with a limited number of steps per each invocation.

**Comment by ruslandoga** [comment]
> ecto_sqlite3 starts [two processes,](https://github.com/ruslandoga/ecto_sqlite/blob/736f7258129a8c2e45c3d869ae960c187fef63ab/lib/ecto/adapters/sqlite/connection.ex#L12-L26) a reader and a writer (for a readwrite db), with reader multiplexing select statements and waiting for transaction close on changes to db, and with writer just queueing writes and Repo.query calls

This doesn't seem to improve much as I'm checking `pragma data_version` before each SELECT which slows down small queries ~x2. And without checking `pragma data_version` before each SELECT we run into the possibility of failing to read own writes. So it seems like multiplexing SELECT statements on the same connection won't work.

**Comment by josevalim** [comment]
> And without checking pragma data_version before each SELECT we run into a possibility of failing to read own writes.

What do you mean, can you expand? I am not well versed on sqlite pragmas. :)

**Comment by ruslandoga** [comment]
Since ongoing selects piggy back on the same read transaction we need to know when it's safe to issue new reads and when it's not. By safety here I mean reading own writes.If we write and then read from the read-only connection that has started the read transaction before our write, the data we wrote would be invisible to us. So to avoid that I'm checking data_version pragma which, if changes since the last invocation, means other connections wrote something into the database. Since I'm doing it before each select, it negatively affects read performance.

**Comment by josevalim** [comment]
For Ecto, I would say transactions always go to the write-based connection. Would that solve this problem?

**Comment by ruslandoga** [comment]
Unfortunately this wouldn't help since selects open a read transaction implicitly if there isn't one already. So the issue is with the concurrent/overlapping selects extending that read transaction and making the new data invisible.

**Comment by josevalim** [comment]
I see, thank you, very helpful. So I guess we have no other option than having a single process that we send everything through?

**Comment by warmwaffles** [comment]
Or perhaps, we just live with the DirtyNIF for now. 😞

**Comment by josevalim** [comment]
I believe we most likely still need DirtyNIFs. It is more of a question of removing the DBConnection dependency. :)

**Comment by josevalim** [comment]
I am reading this once more and I am wondering, how we can improve and simplify things. We have three options:

1. This library only exposes Dirty NIFs. It is up to users to wrap them in processes and guarantee whatever isolation model they care about. The reason why this makes sense is that, even if we provide safe guards here, someone can run two operating system processes and reproduce some of the isolation issues. As long as the NIFs are safe (no segfault, memory leak, etc), we should be good.

2. This library provides a DBConnection wrapper, with serial reads and serial writes, and a single connection. Which is what happens today.

3. This library provides a DBConnection wrapper, with a serial connection that allows concurrent reads and serial writes. We allow multiple reads to run at the same time, as soon as there is a write, we wait for all reads to finish, perform the write, and then start executing the reads again. This should be an improvement over the status quo today (and justifies keeping DBConnection).

Pretty much all other designs we tried were not good enough.

Have I summed up things properly? :)

**Comment by LostKobrakai** [comment]
I don't think the assessment of what is the current state with point 2 is correct: What we currently have with ecto_sqlite3 is DBConnection with concurrent reads and writes. Concurrent writes block at the sqlite level though and are either successfuly serialized with sqlites busy mechanic or some writes run into busy timeouts and crash.

**Comment by josevalim** [comment]
Thank you @LostKobrakai. If that's the case, then we might as well have gone with Dirty NIFs all the way? What would DBConnection gives us if we allow both concurrent reads and writes?

**Comment by LostKobrakai** [comment]
To my knowledge that was done to integrate with ecto (hence it being in ecto_sqlite3 and not this repo). Not sure if there would've been other / simpler options to make ecto work with exqlite.

**Comment by warmwaffles** [comment]
The `DbConnection` implementation is done in the `ecto_sqlite3` project. I intended this library to be as a mechanism by which to interact with sqlite directly. My intention was to keep the ecto implementation away from this repo so that in theory a different sqlite NIF library _could_ replace this should it be necessary.

**Comment by josevalim** [comment]
@warmwaffles I don't think you can replace dirty nifs though. That's the way to do C integration. Your issue with query operations may be addressed by the following:

1. return a NIF resource (a reference) on every query from the NIF
2. pass this NIF resource to read the query result
3. set a destructor to the NIF resource that will cancel the query if it was not cancelled yet

WDYT? Could that work? We are solving similar issues in Apache ADBC.

I would probably still say then that the DBConnection abstraction could be moved to the ecto_sqlite3 repo, it is not really necessary for SQLite3 itself, and then it should be debated if ecto_sqlite3 itself needs DBConnection. :)

And in case I haven't seen it before, thank you for your work and everyone's feedback and inputs on the thread. :heart:

**Comment by warmwaffles** [comment]
That may work. Handing back a handle to cancel a running query and doing that transparently.

> I would probably still say then that the DBConnection abstraction could be moved to the ecto_sqlite3 repo, it is not really necessary for SQLite3 itself, and then it should be debated if ecto_sqlite3 itself needs DBConnection.

I agree with this. I'll start looking at moving it out of this repository soon. Probably over the holiday break I'll knock it out.

**Comment by josevalim** [comment]
> Concurrent writes block at the sqlite level though and are either successfuly serialized with sqlites busy mechanic or some writes run into busy timeouts and crash.

Apparently sqlite does not automatically serializes it for you, you are respnsible to do it, which pretty much means a DBConnection with a pool of 1 is the way to go. So perhaps this can be closed? More info here: https://www.sqlite.org/rescode.html#busy

**Comment by warmwaffles** [comment]
Yea sadly at the current state of things, I don't know if I can interrupt the current executing query to cancel it. I haven't had a whole lot of time to dedicate to exploring a way to timeout/cancel executions.

**Comment by LostKobrakai** [comment]
> Yea sadly at the current state of things, I don't know if I can interrupt the current executing query to cancel it.

https://www.sqlite.org/c3ref/interrupt.html

This seems to be an explicit feature of sqlite. Couldn't this be used?

**Comment by warmwaffles** [comment]
Yes, that's the documentation I was looking at that spurred this ticket.

**Comment by dimitarvp** [comment]
So I have started and stopped writing this comment probably 17 times during the last two weeks. 😃 I hate talking before I have a 100% solution but this time around, and since I finally have inspiration and some time + energy to work on my Elixir<=>Rust SQLite library, and as it is progressing much faster and nicer than I thought (basic feature completeness achieved a few days ago without stuff like sessions, backups, UDFs and the like that I don't find critical for v1), I figured it's time to try and address everything in this thread -- which took me 4 full reads and taking notes like a school student; hey, I never claimed I was smart.

Let me know if I missed something important. Please!

### --- Single library vs. two or more ---
 
I agree with Jose that trying to put everything in a single library presents architectural challenges (too long to outline in this very likely overloaded comment IMO, but ask if you are curious). I already have `xqlite` working quite fine and it covers the "raw" NIF access to SQLite with some very nice guard rails (never aborts / panics) -- though test coverage needs little more work, which I'll put in _VerySoon™_. Next thing planned is an `xqlite_ecto3` library which should be done in several weeks... right after I add a few more features that are extremely important for Ecto / DBConnection integration. Would love critical reviews of it, with me naively assuming you guys will have the time and desire to do so.

### --- Read / write serialization ---

I grumpily acknowledge the problems with read and write serialization, with or without WAL, and how that can be a major pain for the users of the library. I thought a lot about it and can't find a perfect solution. Mine follows:

- `xqlite_ecto3` will integrate closely with DBConnection and `pool_size` will definitely be more than one (unless the `use Ecto.Repo, ...` block specifies 1, of course).
- WAL / WAL2 mode is _hugely_ preferred and I'll make that blindingly obvious in the README file(s).
- Ditto for using `PRAGMA busy_timeout = 5000;` (configurable). I am still mulling over it but I am fairly convinced that I should just transparently put that in place in the Ecto / DBConnection integration layer (i.e. the future `xqlite_ecto3` lib). Without it almost everything in this list becomes rather moot. Is my thinking correct? 🤔 
- My NIF library already specifically catches the busy or locked DB cases and returns a proper error tuple (with atoms) that you can easily pattern-match on if you want to do f.ex. several recursive retries (or you can crash 🤷). My main priority is to give you the choice. Still not sure how would I expose that choice inside an integration layer library _and_ then to the end-user of the library... any ideas? I kind of got stuck here as I _really_ don't want to use `Config` in a library. Might even ask an LLM and laugh at its expense on this topic. I need some reassurance that there's something out there that's stupider than me.

And just to make this crystal clear: my Elixir + Rust code already took care of synchronization. No separate Elixir processes will access the connection handle at the same time -- I fought very hard to avoid a `Mutex` but was ultimately unable to do so because TL;DR the BEAM VM can move an Elixir process to another OS thread at any yield point, and the Rust compiler does not take kindly to that at all.

The `Mutex` lock is per-handle, of course. Which implies SQLite must always be open in "no mutex" mode. I'll make sure this is spelled out in the Rust code, even though me using `rusqlite`'s defaults already does it, it's just implicit.

So yeah, my code will _not_ do database-level locking, ever. I let SQLite do it via its innate one-writer-at-a-time limitation. We handle it via timeouts and cancellations.

### --- Zombie queries / executes ---

I have experimented with cancellation via using SQLite's `set_progress_handler` mechanism. I have, rather non-creatively, called it `CancelToken` (it can be created and cancelled and unregistered). It works very well already! (Not pushed to the repo yet.) I spent an entire afternoon wondering about the perfect value to pass to SQLite (== number of SQLite VM instructions that have to be executed before the progress handler is called [again]) and, some mailings list threads later, I settled at 1000. Which roughly means "on CPU-intensive queries / executes you might get 50 progress handler calls during a single millisecond, yet you might only get called once every 20 milliseconds on disk-heavy queries / executes". Far from what we would call ideal but... researched as much as I could before burning out and for now it's going to be like this.

Crucially, and thanks to Rust's RAII mechanisms, I now have draft code that automatically unregisters an SQLite progress handler either on success or error (tested). Unless I am thinking about this wrong, panics are not interesting to handle because most of any state you would have would be at this point either corrupted or about-to-be-corrupted.

To clarify: I am combining SQLite's `set_progress_handler` with Rust's atomic boolean facilities and have the progress handler check for its value becoming `true` (and that will be done via a NIF function i.e. `cancel(token)`); if that happens then `sqlite3_interrupt` is called. As far as I am aware, that's the way to do it and I've seen people report success doing it similarly.

EDGE CASE: timeout for a query / execute expires but the NIF operation finishes before SQLite calls the progress handler and `sqlite3_interrupt` right after. Gosh, I agonized _hours_ on this but I can't find a way out. If it happens, I suggest we just trust the BEAM VM facilities (i.e. if you use `Task.await` with a timeout) and if a timeout is hit then a timeout is hit and that's that; silently discard that now-outdated success result. Still, this scenario should be extremely rare so this tradeoff should be fine, I believe. Thoughts?

### --- Streaming ---

I would have forgotten about this if I haven't read this thread several times. But yes, streaming is absolutely getting added in the next few days. It's laborious and annoying to add (a lot of Rust boilerplate) but I love the idea and I came to Elixir for stuff like this in the first place i.e. to have lag-less and smooth operation with oft yields. So I feel quite strongly about this and it _is_ getting done (basically through prepared statements lifecycle management). Very soon.


### --- End ---

Again, please let me know if I neglected a worry expressed here. I _really_ want to keep pushing my work further and provide a full-blown SQLite library _and_ as perfect an Ecto / DBConnection integration as at all permitted by both SQLite and the BEAM VM. I want to get this done and for it to be done as well as my limited ability allows me.

@warmwaffles Hope you don't find me talking about a competing library in your own library's issue tracker rude. Just wanted to go where the really important discussion is. I took some lessons from your code, by the way. Nice work.

And cheers all. Loved the discussion. Opened my eyes about a few things.

@josevalim / @LostKobrakai / @ruslandoga
